### PR TITLE
check APIStatus.Code in Is* family of functions

### LINKED ISSUE #101841: apimachinery/errors: IsForbidden does not match Code; it only matches Reason
#### What happened:

Checked a `Create` method error using `"k8s.io/apimachinery/pkg/api/errors".IsForbidden` when the error was caused by a validating webhook that returns a 403 response.

In this particular case, the error was generated from a validating webhook implemented with controller-runtime that returns controller-runtime's `Denied` response.

#### What you expected to happen:
`"k8s.io/apimachinery/pkg/api/errors".IsForbidden(err)` to return true.

#### How to reproduce it (as minimally and precisely as possible):
`"k8s.io/apimachinery/pkg/api/errors".IsForbidden(err)` returned false

#### Anything else we need to know?:

Many (but not all) of the other `Is*` family of functions in `k8s.io/apimachinery/pkg/api/errors` suffer the same kind of problem.

I have changes ready to submit in a PR if we can get agreement that this is a problem and that it should be fixed in the functions that ought to check the `Code` of an erorr value that is or wraps an `"k8s.io/apimachinery/pkg/api/errors".APIStatus`

#### Environment:
- Kubernetes version (use `kubectl version`):
1.19.3 for the server

### PR DESCRIPTION
**Pull request by bhcleek**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Check the Code in the Is family of functions where it makes sense. This
was already done in a couple of functions, but there were many others
where checking the code makes sense.

IsConflict has to check IsAlreadyExists, because IsAlreadyExists uses
http.StatusConflict as the error.

IsTimeout has to check IsServerTimeout, because IsServerTimeout also
uses http.StatusGatewayTimeout.

#### Which issue(s) this PR fixes:
Fixes #101841

#### Does this PR introduce a user-facing change?
Not that I know of, though, it may depend on the definition of user-facing; it expands the checks in api-machinery's error's Is* family of functions to include the http codes.

### CODE REVIEW

**Comment by smarterclayton** [comment]
Hrm.  There was (at one point) a deliberate decision not to do this, but it hasn’t been revisited in four years or so since the last major change.  My primary concern is overriding reason (which would be a breaking change if someone had an empty reason on a non standard aggregated api).

We probably can’t change this without a fair amount of community review and we’d still likely be at risk of breaking callers.   Would like some other opinions but i’m not sure I feel comfortable adding this check without more data.

/hold

**Comment by liggitt** [comment]
I agree that overriding a present `Reason` based on the `Code` value is questionable

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/errors/statuserror.go#L28-L57 is where we translate the status returned by the webhook into the status returned to the user. If a webhook is only returning a code and is not setting `Reason`, I could maybe see populating it in that method

**Comment by liggitt** [comment]
> could we _at_ least patch the documentation here to make it clear that the `Code` is disregarded despite the Kubernetes API documentation

making the documentation accurate is definitely a good idea

> and also patch controller-runtime to provide what's currently being used as the `Reason` as the `Message` instead and use `Reason` values that match the expectations of `k8s.io/apimachinery/pkg/api/errors` based on the `Code`?

that also seems like a good idea, please open an issue at https://github.com/kubernetes-sigs/controller-runtime/issues for that

**Comment by bhcleek** [comment]
@liggit I'll open an issue for controller-runtime shortly. Should I create an issue there before starting work on it?

> making the documentation accurate is definitely a good idea

Can you clarify the state of this PR for me? Are we waiting for @jpbetz to weigh in? 

If this isn't going to be merged, could you give me some understanding of how bugs can be resolved when apimachinery's behavior does not match Kubernetes' API and how apimachinery relates to the documented API (I'm happy to go read docs if you can point me to the right docs that specify that apimachinery shouldn't be expected to conform to the documented Kubernetes API)?

**Comment by liggitt** [comment]
> If this isn't going to be merged, could you give me some understanding of how bugs can be resolved when apimachinery's behavior does not match Kubernetes' API and how apimachinery relates to the documented API (I'm happy to go read docs if you can point me to the right docs that specify that apimachinery shouldn't be expected to conform to the documented Kubernetes API)?

The impact on existing callers of these functions of changing these functions' logic is unknown when dealing with errors that did not originate from the Kube API server (generic http errors).

I don't know of a good way to 1) bring the implementation in line with docs and 2) guarantee this doesn't adversely affect clients already using these methods.

If the main way this was encountered when talking to a Kubernetes API server was from a webhook that did not fill out the Reason field, setting that based on the code seems like a much more scoped change we could start with.

**Comment by bhcleek** [comment]
Thanks for getting back to me @liggitt 

### regarding these changes in particular

> The impact on existing callers of these functions of changing these functions' logic is unknown when dealing with errors that did not originate from the Kube API server (generic http errors).

This should not cause generic HTTP errors to considered. This PR _only_ checks those errors that satisfy the `"k8s.io/apimachinery/pkg/api/errors".Status` interface, so only `"k8s.io/apimachinery/pkg/apis/meta/v1".Status` values will be considered. In such a case the responder clearly meant for the `Status` value to be used and the `Status` documentation says that the `Code` is authoritative and the other fields are informational only.

> I don't know of a good way to 1) bring the implementation in line with docs and 2) guarantee this doesn't adversely affect clients already using these methods.

In this case, the API clearly documents the expected use of `Code` and the `k8s.io/apimachinery/pkg/api/errors` looks like the right thing to check. Notably, this PR is additive; it doesn't remove the checks on the `Reason`, it only _also_ checks the `Code`. Using `IsForbidden` as an example, code that was returning true from `IsForbidden` will still return true. In cases where a webhook is returning the right `Code`, but not the expected `Reason` values, `IsForbidden` will also now correctly return true in that case.

> If the main way this was encountered when talking to a Kubernetes API server was from a webhook that did not fill out the Reason field, setting that based on the code seems like a much more scoped change we could start with.

I created https://github.com/kubernetes-sigs/controller-runtime/pull/1539 to address this concern on the controller-runtime side.

### larger questions from a new contributor
> The impact on existing callers of these functions of changing these functions' logic is unknown when dealing with errors that did not originate from the Kube API server (generic http errors).
> I don't know of a good way to 1) bring the implementation in line with docs and 2) guarantee this doesn't adversely affect clients already using these methods.

I think this is where my confusion comes in. Resolving bugs will always cause some change in behavior. As a new contributor, what principle should I apply in the future when deciding whether to pick up a Kubernetes issue in order to avoid doing work that a priori won't be approved? This is likely a much larger topic; where is the appropriate place to discuss it (e.g. the apimachinery sig)?

**Comment by bhcleek** [comment]
@liggitt I know you're super busy, but it was suggested to me by some other k8s maintainers that a gentle nudge here might be helpful in case this has gotten lost in the shuffle.

**Comment by liggitt** [comment]
apologies for the delay, this is on my list to pick back up now that the 1.22 crush is ~over

**Comment by liggitt** [comment]
ok, finally got back to this.

I see the following scenarios:

1. status errors parsed from generic http responses (done in `k8s.io/apimachinery/pkg/api/errors#NewGenericServerResponse`)
2. status errors parsed from responses formatted as `meta.Status`
    1. which set `reason` to a known reason
        1. and set `code` to match
        2. and set `code` to a mismatched value (or do not set `code`)
    1. which set `reason` to an unknown reason (or do not set `reason`)
        1. and set `code` to a known code
        2. and set `code` to an unknown code (or do not set `code`)

I'm using the following definitions:
* "known reason" == reason defined in `k8s.io/apimachinery/pkg/apis/meta/v1` / reason for which one of the `k8s.io/apimachinery/pkg/api/errors#Is<ErrorType>(error)` helper functions would return `true`
* "known http status code" == status code mapped to known reasons in `k8s.io/apimachinery/pkg/api/errors#NewGenericServerResponse`

Scenario | Current outcome | Outcome with this PR
---|---|---
1   | good | unchanged
2.i.a | good | unchanged
2.i.b | helpers honor reason | helpers honor reason _or_ code - this PR changes behavior so two `Is<ErrorType>(error)` functions could both return true for the same error, one based on `reason` and one based on `code`
2.ii.a | helpers return false | helpers honor code
2.ii.b | helpers return false | unchanged

The only scenario that gives me pause is 2.i.b. I think it is confusing for two `Is<ErrorType>` functions to both return true on the same error.

One way around this would be to limit the new fallback to `code` to inputs where `reason` is empty/unknown.

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:481]: Rather than adding a helper just to extract code, and type-extracting every error twice, could do both in a single step. Combined with the suggestion to only fall back to code for unrecognized reasons, could look like this:

```go
reason, code := ReasonAndCodeForError(err)
if reason == metav1.StatusReasonNotFound {
  return true
}
if !knownReasons[reason] && code == http.StatusNotFound {
  return true
}
return false
```

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:481]: that approach would also make the `IsConflict` change consistent... since `StatusReasonAlreadyExists` would not pass the `if !knownReasons[reason]` test

**Comment by bhcleek** [comment]
@liggit thank you for taking the time to respond 🙇 

I'll make those changes ASAP

**Comment by bhcleek** [comment]
@liggitt PTAL

I think I made the changes you're looking for.

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:710]: since this was already checking code without regard for reason, continue doing so and add a comment noting why this is different than the other helpers

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:724]: since this was already checking code without regard for reason, continue doing so and add a comment noting why this is different than the other helpers

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:667]: StatusGatewayTimeout is used for this in NewGenericServerResponse and NewTimeoutError, not StatusRequestTimeout

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:681]: StatusInternalServerError is used for this in NewServerTimeout, not StatusGatewayTimeout

**Review by liggitt — COMMENTED** [review]
this needs unit tests added for a few of the noted scenarios (not exhaustively, but exercising the known/unknown reason/code combinations for a few of the methods)

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (bhcleek) [reply staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:667]: I'm not sure I understand the significance of `NewGenericServerResponse` and `NewTimeoutError` as it relates to this change; if those methods were used, then the reason would be known, and this if block wouldn't be executed, but that doesn't even matter, because this code won't even be executed because both `NewGenericServerResponse` and `NewTimeoutError` will set the reason to `StatusReasonTimeout`.

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (bhcleek) [reply staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:681]: I'm not sure I understand the significance of `NewServerTimeout` as it relates to this change; if that method were used, then the reason would be known, and this if block wouldn't be executed, but that doesn't matter, because this code won't be executed because `NewServerTimeout` will set the reason to `StatusReasonServerTimeout`.

**Comment by bhcleek** [comment]
@liggit I've added the tests and comments you asked for. I'll add the other changes you mentioned if you still think it's necessary after you've seen my response in the threads of your last review.

**Comment by bhcleek** [comment]
@liggit I spelled your username incorrectly when I first posted my last comment. I'm not sure if GitHub would notify you when I corrected the spelling. Apologies if this is a double ping.

**Comment by bhcleek** [comment]
@liggitt I spelled your username incorrectly when I first posted my last comment. I'm not sure if GitHub would notify you when I corrected the spelling. Apologies if this is a double ping.

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:681]: My understanding is that code == http.StatusGatewayTimeout should make `IsTimeout` return true, not this method.

This method is scoped to retryable timeouts, and per these docs there is no http code specific to that, so adding an http code fallback here doesn't seem appropriate:

https://github.com/kubernetes/kubernetes/blob/release-1.21/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L799-L806

**Comment by bhcleek** [comment]
@liggitt thank you for your explanations today. I've pushed changes in accordance with your explanations in the last commit.

**Comment by liggitt** [comment]
/ok-to-test
/sig api-machinery
/triage accepted
/priority backlog

Go ahead and squash to a single commit, then this lgtm. Will give @smarterclayton a chance to review where this ended up, then unhold next week if there's no further concerns.

File: staging/src/k8s.io/apimachinery/pkg/api/errors/errors_test.go
Comment (MorrisLaw) [inline staging/src/k8s.io/apimachinery/pkg/api/errors/errors_test.go:540]: nit: Does `_any_` actually show up as _any_ anywhere? Or can we change it to `any`?

**Review by MorrisLaw — APPROVED** [review]
left a small nit, but lgtm! Thanks Billie :+1:

**Comment by pacoxu** [comment]
/sig api-machinery

The bot goes wrong.

**Pull request merged**
